refactor, kernel: Decouple ArgsManager from blockstorage

sedited opened the pull request:
The libbitcoin_kernel library should not rely on the `ArgsManager`, but rather use option structs that can be passed to the various classes it uses. This PR removes reliance on the `ArgsManager` from the `blockstorage.*` files. Like similar prior work, it uses the options struct in the `BlockManager` that can be populated with `ArgsManager` values.

Some related prior work: https://github.com/bitcoin/bitcoin/pull/26889 https://github.com/bitcoin/bitcoin/pull/25862 https://github.com/bitcoin/bitcoin/pull/25527 https://github.com/bitcoin/bitcoin/pull/25487

Related PR removing blockstorage globals: https://github.com/bitcoin/bitcoin/pull/25781

sedited committed:
refactor, BlockManager: Replace fastprune from arg with options

Remove access to the global gArgs for the fastprune argument and
replace it by adding a field to the existing BlockManager Options
struct.

When running `clang-tidy-diff` on this commit, there is a diagnostic
error: `unknown type name 'uint64_t' [clang-diagnostic-error] uint64_t
prune_target{0};`, which is fixed by including cstdint.

This should eventually allow users of the BlockManager to not rely on
the global gArgs and instead pass in their own options.

sedited committed:
refactor: Move functions to BlockManager methods

This is a commit in preparation for the next few commits. The functions
are moved to methods to avoid their re-declaration for the purpose of
passing in BlockManager options.

The functions that were now moved into the BlockManager should no longer
use the params as an argument, but instead use the member variable.

In the moved ReadBlockFromDisk and UndoReadFromDisk, change
the function signature to accept a reference to a CBlockIndex instead of
a raw pointer. The pointer is expected to be non-null, so reflect that
in the type.

To allow for the move of functions to BlockManager methods all call
sites require an instantiated BlockManager, or a callback to one.

sedited committed:
refactor, blockstorage: Replace blocksdir arg

Add a blocks_dir field to the BlockManager options. Move functions
relying on the global gArgs to get the blocks_dir into the BlockManager
class.

This should eventually allow users of the BlockManager to not rely on
the global Args and instead pass in their own options.

sedited committed:
refactor, blockstorage: Replace stopafterblockimport arg

Add a stop_after_block_import field to the BlockManager options. Use
this field instead of the global gArgs.

This should allow users of the BlockManager to not rely on the global
Args.

fanquake commented:
```bash
validation.cpp:80:13: error: using decl 'ReadBlockFromDisk' is unused [misc-unused-using-decls,-warnings-as-errors]
using node::ReadBlockFromDisk;
            ^
/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/validation.cpp:80:13: note: remove the using
using node::ReadBlockFromDisk;
```

TheCharlatan commented:
> ```shell
> validation.cpp:80:13: error: using decl 'ReadBlockFromDisk' is unused [misc-unused-using-decls,-warnings-as-errors]
> using node::ReadBlockFromDisk;
>             ^
> /tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/validation.cpp:80:13: note: remove the using
> using node::ReadBlockFromDisk;
> ```

Thank you, I'm improving my clang tidy workflow.

brunoerg reviewed (commented):
Concept ACK

brunoerg commented on src/bitcoin-chainstate.cpp:
```diff
@@ -85,7 +85,8 @@ int main(int argc, char* argv[])
     };
-    ChainstateManager chainman{chainman_opts};
+    const node::BlockManager::Options blockman_opts{};
+    ChainstateManager chainman{chainman_opts, blockman_opts};
```
nit: Since `blockman_opts` is empty here, I think you can pass it directly instead of creating `blockman_opts` for it.
```cpp
ChainstateManager chainman{chainman_opts, {}};
```

brunoerg commented on src/net_processing.cpp:
```diff
@@ -3869,7 +3866,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             if (pindex->nHeight >= m_chainman.ActiveChain().Height() - MAX_BLOCKTXN_DEPTH) {
                 CBlock block;
-                bool ret = ReadBlockFromDisk(block, pindex, m_chainparams.GetConsensus());
+                bool ret = m_chainman.m_blockman.ReadBlockFromDisk(block, pindex, m_chainparams.GetConsensus());
```
nit:
```cpp
const bool ret{m_chainman.m_blockman.ReadBlockFromDisk(block, pindex, m_chainparams.GetConsensus())};
```

TheCharlatan commented:
Updated e8b095f8ad65544f18ad69201d95e11ddf689f1f -> 899ba71e9d7caa370d8a0c4bed2e514f51024dc6 ([removeBlockstorageArgs_0](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_0) -> [removeBlockstorageArgs_1](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_1), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_0..removeBlockstorageArgs_1)) with suggested changes. 
Rebased 899ba71e9d7caa370d8a0c4bed2e514f51024dc6 -> b640db3c5fbe49af58dbc9087a96cbb493fa9574 to get recent PR's with kernel build improvements.

mzumsande commented on src/kernel/blockmanager_opts.h:
```diff
@@ -0,0 +1,19 @@
+namespace kernel {
+
+/**
+ * An options struct for `BlockManager`, more ergnomically referred to as
```
ergonomically

TheCharlatan commented:
Updated b640db3c5fbe49af58dbc9087a96cbb493fa9574 -> b640db3c5fbe49af58dbc9087a96cbb493fa9574 ([removeBlockstorageArgs_2](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_2) -> [removeBlockstorageArgs_3](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_3), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_2..removeBlockstorageArgs_3)) with suggested typo fix.

mzumsande reviewed (commented):
Concept ACK

Seeing all that effort to pass around the hacky `fastprune` arg (which doesn't make any sense outside of functional tests) seems a bit sad, so I wonder if it's really inevitable.

Did you consider whether it's possible to also move `BlockFileSeq` into blockmanager? I guess it doesn't work just because of one remaining call site for `node::ReadBlockFromDisk` from `zmqpublishnotifier`?

dergoegge commented:
Concept ACK

TheCharlatan commented:
Thanks for the reviews!

> Seeing all that effort to pass around the hacky fastprune arg (which doesn't make any sense outside of functional tests) seems a bit sad, so I wonder if it's really inevitable.
> Did you consider whether it's possible to also move BlockFileSeq into blockmanager? I guess it doesn't work just because of one remaining call site for node::ReadBlockFromDisk from zmqpublishnotifier?

Yes, totally echo this sentiment. Indeed BlockFileSeq and its dependants are not class members because of the call site in `zmqpublishnotifier`. I refrained from further refactoring because of this. If somebody has an idea on how to handle the zmq case I'll gladly rework this pull request.

Updated 3ded321b878e7ad2bdbd9d542ffb2e36428396ec ->  596404317bef425179d68009a76ac185db4da9a4 ([removeBlockstorageArgs_3](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_3) -> [removeBlockstorageArgs_4](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_4), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_3..removeBlockstorageArgs_4)) removing unused declaration.
Rebased 596404317bef425179d68009a76ac185db4da9a4 ->88e88a598581f8a081838484ca6584d96e00f2e0 ([removeBlockstorageArgs_4](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_4) -> [removeBlockstorageArgs_5](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_5), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_4..removeBlockstorageArgs_5)) to resolve conflicts.

TheCharlatan commented:
Updated  88e88a5 -> f87c39399823e22c553b797cc66fa4063462a32b ([removeBlockstorageArgs_5](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_5) -> [removeBlockstorageArgs_6](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_6), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_5..removeBlockstorageArgs_6)) with linter fix.

TheCharlatan commented:
Rebased 6d9826f182d9122d4464d35d6682dc6fb4b1116e -> 593a9f36609a74f47175e681a3921f3975272766 ([removeBlockstorageArgs_7](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_7) -> [removeBlockstorageArgs_8](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_8), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_7..removeBlockstorageArgs_8)) on newest master to resolve conflicts.

sedited committed:
refactor/iwyu: Complete includes for blockmanager_args

TheCharlatan commented on src/zmq/zmqpublishnotifier.cpp:
```diff
@@ -248,9 +253,12 @@ bool CZMQPublishRawBlockNotifier::NotifyBlock(const CBlockIndex *pindex)
     const Consensus::Params& consensusParams = Params().GetConsensus();
+    BlockManagerOpts blockman_opts{};
+    Assert(!ApplyArgsManOptions(gArgs, blockman_opts)); // already checked on init
+    BlockManager blockman{blockman_opts};
```
I may have missed some nuance here. To me it seemed acceptable, because functions are protected by `cs_main`. For const functions like`ReadBlockFromDisk` this seems safe to me (though I'm still new to the locking model and might be missing something here). However doing block write and pruning operations with two blockmanagers is fragile, since their internal state might no longer reflect the state on disk. 

I'm not sure how to proceed here. I could mark the `BlockManager` here as const, that would at least ensure that it's state cannot change.

fanquake replied:
> I may have missed some nuance here. To me it seemed acceptable, because functions are protected by cs_main. For const functions likeReadBlockFromDisk this seems safe to me

Note that according to the PR description in #27006 there is currently at least one `cs_main` related bug in `ReadBlockFromDisk`.

TheCharlatan replied:
> I feel this though, I'll try out the first and second option you laid out.

Both are branching out a bit too much in my taste for the refactor we seek to achieve here. If the `const BlockManager` is a non-starter, I can always revert back to https://github.com/bitcoin/bitcoin/commit/f87c39399823e22c553b797cc66fa4063462a32b

TheCharlatan replied:
This patch is much simpler than what I came up with, thanks a lot!

TheCharlatan commented on src/node/blockstorage.h:
```diff
@@ -236,11 +240,11 @@ class BlockManager
     bool ReadBlockFromDisk(CBlock& block, const FlatFilePos& pos, const Consensus::Params& consensusParams) const;
     bool ReadBlockFromDisk(CBlock& block, const CBlockIndex* pindex, const Consensus::Params& consensusParams) const;
```
Thank you for the suggestion, will split this up.

ryanofsky replied:
In commit "refactor: Move functions to BlockManager methods" (1def580e80a1892e473324016db9a0f2ffec0c27)

re: https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1140490220

> Thank you for the suggestion, will split this up.

Seems like this could be marked resolved

TheCharlatan commented:
Updated 54d8644dce180197fa657e9b68d6de63cf4c8072 -> 9f5600742c65a2421c9fe5a2a2670e86a25ef696 ([removeBlockstorageArgs_9](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_9) -> [removeBlockstorageArgs_10](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_10), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_9..removeBlockstorageArgs_10)) to split a commit moving methods and instantiating a block manager in zmq, and replacing the fastprune arg. I also added a short inline comment explaining the zmq's blockmanager const declaration and threading requirements.

TheCharlatan commented:
Thank you for the discussion and suggestions.

Updated 9f5600742c65a2421c9fe5a2a2670e86a25ef696 -> 97bf70119e5b8567bcdc553d59b18a09023fd05a ([removeBlockstorageArgs_10](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_10) -> [removeBlockstorageArgs_11](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_11), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_10..removeBlockstorageArgs_11)) to add @dergoegge's patch as suggested in his [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1146221073). With this latest push, there is no additional `BlockManager` being instantiated anymore.

TheCharlatan commented:
Not sure why the functional tests are failing. The test failing on the Win64 job is passing on the 32-bit job and vice-versa. Both are passing locally.

ryanofsky commented:
Win64 wallet_create_tx.py failure is probably fixed by #27318

sedited committed:
refactor: Declare g_zmq_notification_interface as unique_ptr

Ensures better memory safety for this global. This came up during
discussion of the following commit, but is not strictly required for its
implementation.

TheCharlatan commented:
Thank you for the review @mzumsande

Updated 97bf70119e5b8567bcdc553d59b18a09023fd05a -> 09f950e1f3608e3019e97236969b152e9252a220 ([removeBlockstorageArgs_11](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_11) -> [removeBlockstorageArgs_12](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_12), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_11..removeBlockstorageArgs_12)) 
* Addressed @mzumsande's [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1160180267) by making the `g_zmq_notification_interface` a `std::unique_ptr` and checking that it is not already the owner of a `CZMQNotificationInterface` object before assigning to it. 
* Pass a reference of the node context to the `CZMQNotificationInterface` to ensure that we access the currently initialized `BlockManager` within the zmq code. In the previous revision, the reference to the `BlockManager` might have been pointing to a destructed object.

TheCharlatan commented:
Rebased 908d95e46f15f6b4def9534bc09fcdb3c0a5ef54 -> 1a7d73d30e312e862be4560601302e79725b1bc7 ([removeBlockstorageArgs_13](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_13) -> [removeBlockstorageArgs_14](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_14), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_13..removeBlockstorageArgs_14))

dergoegge commented on src/zmq/zmqpublishnotifier.cpp:
```diff
@@ -250,7 +249,7 @@ bool CZMQPublishRawBlockNotifier::NotifyBlock(const CBlockIndex *pindex)
     CDataStream ss(SER_NETWORK, PROTOCOL_VERSION | RPCSerializationFlags());
     CBlock block;
-    if (!ReadBlockFromDisk(block, pindex, consensusParams)) {
+    if (!m_node.GetBlockManager().ReadBlockFromDisk(block, pindex, consensusParams)) {
```
nit: Seems unnecessary to pass the entire `NodeContext` when you only need the blockman.

Feel free to ignore, but would be interested in why you chose to do it this way.

TheCharlatan replied:
The blockman and chainstate instances may be deleted during the lifetime of the zmq instance similarly as described in mzsumsande's [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1160180267) when iterating the for loop. If we pass a reference to the `NodeContext` instead, we have better guarantees that its blockman reference points to an actual object.

dergoegge replied:
Is that actually possible? Seems like all the validation interfaces including the zmq one are unregistered before chainman is destroyed?

So it can't be holding a reference to blockman that is actually destroyed, see:

https://github.com/bitcoin/bitcoin/blob/f3f5c9712670fc2c9fb3b57580146aa248a5d36f/src/init.cpp#L327-L343

Or are you just saying that having a reference to the node context has better lifetime guarantees because basically nothing outlives the node context?

TheCharlatan replied:
> Is that actually possible? 

Likely that I am missing something in the logic. The way I see it each iteration of the for loop creates and destroys the `chainman` by calling `std::make_unique` again, while the `Shutdown` code is not called during each iteration.

dergoegge reviewed (commented):
lgtm once the last `gArgs` access is removed.

TheCharlatan commented on src/node/blockstorage.cpp:
```diff
@@ -583,28 +578,28 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune)
+FlatFileSeq BlockManager::BlockFileSeq() const
 {
-    return FlatFileSeq(gArgs.GetBlocksDirPath(), "blk", gArgs.GetBoolArg("-fastprune", false) ? 0x4000 /* 16kb */ : BLOCKFILE_CHUNK_SIZE);
+    return FlatFileSeq(gArgs.GetBlocksDirPath(), "blk", m_opts.fast_prune ? 0x4000 /* 16kb */ : BLOCKFILE_CHUNK_SIZE);
```
Good catch, thank you! Will also remove the `<args.h>` include.

TheCharlatan commented:
Thank you for the review @dergoegge,

Updated 1a7d73d30e312e862be4560601302e79725b1bc7 -> ed3159e0eb105d9f46659bd8f8b2db27d94841de ([removeBlockstorageArgs_14](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_14) -> [removeBlockstorageArgs_15](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_15), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_14..removeBlockstorageArgs_15)) 
* Addressed @dergoegge's [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1173878733), removing a left over `gArgs` reference in `blockstorage.cpp`.

dergoegge commented:
Code review ACK bab12a6a029bed86ddcc15d42a236a855e277e74

dergoegge commented:
Code review ACK ed3159e0eb105d9f46659bd8f8b2db27d94841de

ryanofsky commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -39,12 +43,14 @@ std::list<const CZMQAbstractNotifier*> CZMQNotificationInterface::GetActiveNotif
 }
 
-CZMQNotificationInterface* CZMQNotificationInterface::Create()
+std::unique_ptr<CZMQNotificationInterface> CZMQNotificationInterface::Create(const node::NodeContext& node)
```
Just talked about this offline, but if possible to make ZMQ just take a pointer to BlockManager instead of NodeContext, that would be good. This way ZMQ would only have access to the state it actually needs

TheCharlatan commented:
Thank you for the discussions,

Updated ed3159e0eb105d9f46659bd8f8b2db27d94841de -> 9c197181803e08e1cce4460190cc06fc2e093a80 ([removeBlockstorageArgs_15](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_15) -> [removeBlockstorageArgs_16](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_16), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_15..removeBlockstorageArgs_16)) 
* Reverted change made for passing a `NodeContext` to `ZmqNotificationInterface` back to passing a reference of a `BlockManager`
* Introduced a change declaring the `BlockManager` outside of the `ChainstateManager` and making the `NodeContext` its new owner. References to this `BlockManager` are now passed to declare `ChainstateManager` and `ZmqNotificationInterface`. This could also allow the kernel more control over how blocks are stored.

Rebased 9c197181803e08e1cce4460190cc06fc2e093a80 -> f59f0c91acb7a35b767bb0dc75ed8b10add81d9f ([removeBlockstorageArgs_16](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_16) -> [removeBlockstorageArgs_17](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_17), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_16..removeBlockstorageArgs_17))

dergoegge commented:
re-ACK f59f0c91acb7a35b767bb0dc75ed8b10add81d9f

maflcko reviewed (commented):
f59f0c91acb7a35b767bb0dc75ed8b10add81d9f 🙍

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: f59f0c91acb7a35b767bb0dc75ed8b10add81d9f 🙍
QCpVzQy7eIdu/y9fKVeQ1TjvngAc8zmwi0x9FvX6ykCtM3q58Vmz49PAujSNjQz5zj6bjCMXWAXoHeOH/w6VAQ==
```

</details>

mzumsande commented on src/node/blockstorage.cpp:
```diff
@@ -584,7 +584,7 @@ void BlockManager::UnlinkPrunedFiles(const std::set<int>& setFilesToPrune) const
 FlatFileSeq BlockManager::BlockFileSeq() const
 {
-    return FlatFileSeq(gArgs.GetBlocksDirPath(), "blk", gArgs.GetBoolArg("-fastprune", false) ? 0x4000 /* 16kb */ : BLOCKFILE_CHUNK_SIZE);
+    return FlatFileSeq(gArgs.GetBlocksDirPath(), "blk", m_opts.fast_prune ? 0x4000 /* 16kb */ : BLOCKFILE_CHUNK_SIZE);
```
It wasn't possible to do that in https://github.com/bitcoin/bitcoin/commit/7e0e84c1a3587f8b2ffac00f647d1e2d17febb91 because `BlockFileSeq` wasn't part of `BlockManager` then. One could rearrange the commits though and have 7e0e84c1a3587f8b2ffac00f647d1e2d17febb91 after the commit that moves stuff into BlockManager.

TheCharlatan commented on src/init.cpp:
```diff
@@ -1439,9 +1440,12 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     Assert(!ApplyArgsManOptions(args, blockman_opts)); // no error can happen, already checked in AppInitParameterInteraction
 
+    assert(!node.blockman);
+    node.blockman = std::make_unique<BlockManager>(blockman_opts);
```
In the context of this commit it is necessary to move the assignment out of the for loop, so we don't destruct the `node.blockman` on each iteration and end up with a dangling reference to the original object in zmq. What I did not think about though is that we are also no longer resetting the internal state of the `BlockManager`when the `ChainstateManager` is destructed. So no, I'm no longer sure that this is safe.

TheCharlatan replied:
Thank you for the context, I see that I am conceptually going backwards here.

TheCharlatan replied:
> I wonder if the for loop can be removed, and replaced by the shutdown workflow from https://github.com/bitcoin/bitcoin/pull/26674#issuecomment-1345237799

That does sound more sane than continuously re-indexing forever, or waiting for the user to interrupt. 

The change I pushed will be reverted. How about instead of moving the previously `static` functions of `blockstorage.cpp` into its `BlockManager` class, moving them to a stateless sub-class of the `BlockManager`? Then we can have the node context be the owner of this class, and pass it through as a const reference to the `ChainstateManager` and its `BlockManager`, as well as the `zmq` code, which can now use this class without taking ownership of the entire `BlockManager`?

TheCharlatan replied:
I was unclear with the word "stateless". What I meant was, "does not retain memory of previous interactions", e.g. we set all the options (including the consensus params) on initialization.

TheCharlatan commented on src/init.cpp:
```diff
@@ -1431,7 +1431,9 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
-    BlockManager::Options blockman_opts{};
+    BlockManager::Options blockman_opts{
+        .blocks_dir = args.GetBlocksDirPath(),
+    };
```
I copied this behavior from `ChainstateManagerOpts`. We expect the default constructed `Options` to contain defaults. If we don't have defaults, we assign on object construction. So, yes, in both instances a `const` qualifier (and maybe in others using the same pattern) should be added.

sedited committed:
zmq: Pass lambda to zmq's ZMQPublishRawBlockNotifier

The lambda captures a reference to the chainman unique_ptr to retrieve
block data. An assert is added on the chainman to ensure that the lambda
is not used while the chainman is uninitialized.

This is done in preparation for the following commits where blockstorage
functions are made BlockManager methods.

TheCharlatan commented on src/node/blockmanager_args.cpp:
```diff
@@ -5,8 +5,15 @@
 #include <validation.h>
 
+#include <cstdint>
+#include <memory>
```
Seems like it's a bug: https://github.com/include-what-you-use/include-what-you-use/issues/648

Since it sounds like it's going to be fixed in an upcoming version, I'll remove the `memory` include.

TheCharlatan replied:
Huh, yeah, seems like it. But I guess the issue is similar to what is happening here?

hebasto commented:
Concept ACK.

TheCharlatan commented:
Thank you for the review and discussion @MarcoFalke!

Updated f59f0c91acb7a35b767bb0dc75ed8b10add81d9f -> db668b3644883c07064b46b4e2cfd269ac9dffbd ([removeBlockstorageArgs_17](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_17) -> [removeBlockstorageArgs_18](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_18), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_17..removeBlockstorageArgs_18)) 
* Reverted a previous change moving `BlockManager` initialization out of the `ChainstateManager` as discssed in https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1183537074
* Added a commit a614ac59cd753c7ef83899602e0176654c5fff87 introducing an interface function for accessing the `BlockManager::ReadBlockFromDisk`. This was introduced since existing functions were more complex and less efficient than just a direct call. If this interface should not grow, I can drop this commit again and use a different method, like the previously discussed `FindBlock` in the following commit.
* Added a commit 6b9ba7b54097bb32d3c1596b7fdcec4428c27a3a capturing this introduced interface function in a lambda and using it in the zmq code. Capturing the `node.chain` should be safe for the purpose of zmq notifying about block tip changes, even with the for loop re-declaring the `ChainstateManager`.
* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1183565479), reordering the commits. Functions are now moved to `BlockManager` methods before removing `ArgsManager` usage.
* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1183570901), adding `nodiscard` qualifier.
* Addressed @MarcoFalke [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1183569789), adding `const` qualifier to Options.
* Addressed @MarcoFalke [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1183527429), using `m_args` for the test arguments.
* Addressed @MarcoFalke [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1183519088), removing `memory` include.
* No longer use `nullptr`for the `g_zmq_notification_interface` in 8bcd56f2aee46ccc28354882009fc38925e629df

hebasto commented on src/interfaces/chain.h:
```diff
@@ -171,6 +172,9 @@ class Chain
         const FoundBlock& ancestor_out={}) = 0;
 
+    //! Get a CBlock by its index and return whether successful.
+    virtual bool getBlockByIndex(CBlock& block, const CBlockIndex* pindex) = 0;
```
a614ac59cd753c7ef83899602e0176654c5fff87, nit, naming:
```suggestion
    virtual bool getBlockByIndex(CBlock& block, const CBlockIndex* index) = 0;
```

hebasto replied:
No, it is not wrong. Just mentioned it as in new code we do not make a reference to a type of the variable in its name:
```
git grep -e "CBlockIndex\* index"
```

hebasto commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -39,12 +39,14 @@ std::list<const CZMQAbstractNotifier*> CZMQNotificationInterface::GetActiveNotif
     factories["pubhashblock"] = CZMQAbstractNotifier::Create<CZMQPublishHashBlockNotifier>;
     factories["pubhashtx"] = CZMQAbstractNotifier::Create<CZMQPublishHashTransactionNotifier>;
-    factories["pubrawblock"] = CZMQAbstractNotifier::Create<CZMQPublishRawBlockNotifier>;
+    factories["pubrawblock"] = [&get_block_by_index = get_block_by_index]() -> std::unique_ptr<CZMQAbstractNotifier> {
```
6b9ba7b54097bb32d3c1596b7fdcec4428c27a3a

Isn't capturing `[&get_block_by_index]` enough?

hebasto reviewed (commented):
Approach ACK db668b3644883c07064b46b4e2cfd269ac9dffbd.

Changes in `test/util/common.cpp` do not seem required in the dbd739d54386dcd5dd1d945423bc82362263d4e8 _"refactor, BlockManager: Replace fastprune from arg with options"_ commit. Maybe combine them into the following commit?

TheCharlatan commented:
Thank you for the review @hebasto 

Updated db668b3644883c07064b46b4e2cfd269ac9dffbd -> 9033dc6827cc623f1f7176fde120229f36ff5e74 ([removeBlockstorageArgs_18](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_18) -> [removeBlockstorageArgs_19](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_19), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_18..removeBlockstorageArgs_19)) 

* Addressed @hebasto's [comment](https://github.com/bitcoin/bitcoin/pull/27125#pullrequestreview-1413008558), removing unneeded change in a commit, and implementing in the commit it was actually required.
* Addressed @hebasto's [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1184962143), dropping the `p` naming prefix.
* Addressed @hebasto's [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1184963724), removing capture assignment that was accidentally carried over from a previous code iteration.

TheCharlatan commented on src/interfaces/chain.h:
```diff
@@ -171,6 +172,9 @@ class Chain
         const FoundBlock& ancestor_out={}) = 0;
 
+    //! Get a CBlock by its index and return whether successful.
+    virtual bool getBlockByIndex(CBlock& block, const CBlockIndex* index) = 0;
```
Thanks for taking a look. It's also not used by an actual client of the interface, so I think I'll just drop the commit then. No need to needlessly pollute the interface.

TheCharlatan commented:
Rebased e056d6f758382d3418682095ab02f8d487aa641f -> 3b34ac7465919b968795063995f6610a73aa2d29 ([removeBlockstorageArgs_20](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_20) -> [removeBlockstorageArgs_21](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_21), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_20..removeBlockstorageArgs_21))
* Fixed conflicts with #27570
* Added a commit to remove the params argument from the functions moved into the `BlockManager`

maflcko reviewed (approved):
nice ACK 3b34ac7465919b968795063995f6610a73aa2d29 🗣

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: nice ACK 3b34ac7465919b968795063995f6610a73aa2d29 🗣
qW6Kq7M3ZBVaQG1TZ3KYvAfs/o/BxahlxqncOR+LkhBejB6WLxUPgOSJbRR/cxW8UxRJBMDqkIVde9krgNPqCw==
```

</details>

ryanofsky commented on src/index/base.cpp:
```diff
@@ -205,7 +203,7 @@ void BaseIndex::ThreadSync()
 
             CBlock block;
             interfaces::BlockInfo block_info = kernel::MakeBlockInfo(pindex);
-            if (!m_chainstate->m_blockman.ReadBlockFromDisk(block, pindex, consensus_params)) {
```
In commit "refactor: Use BlockManager options for params" (0400fb7b55415867d9ff0aa88898920c60b4b2df)

Seem like you could squash this commit into the last commit and decrease review burden because the changes are mechanical and basically every line that changed here changed last commit. Not important though.

ryanofsky reviewed (approved):
Code review ACK 3b34ac7465919b968795063995f6610a73aa2d29. Left some comments but this looks good as is and would not change unless necessary

TheCharlatan commented on src/init.cpp:
```diff
@@ -1035,8 +1036,9 @@ bool AppInitParameterInteraction(const ArgsManager& args, bool use_syscall_sandb
-        node::BlockManager::Options blockman_opts_dummy{
+        BlockManager::Options blockman_opts_dummy{
             .chainparams = chainman_opts_dummy.chainparams,
+            .blocks_dir = args.GetBlocksDirPath(),
```
> This is a slightly different approach to -fastprune and others, which are set in ApplyArgsManOptions().
So I wonder what is the criterion for an arg to be set in ApplyArgsManOptions(), as opposed to being initialized directly? 

Initializing in the list is done for option members that don't have sane default or empty values. This also has precedent in the `ChainstateManager::Options`. 

> And, closely connected, isn't the current approach of not calling ApplyArgsManOptions() in some spots (setup_common, some unit tests) a bit brittle?

Yes, the current situation does not seem to be ideal. Sometimes these `extra_args` that the user can supply are currently parsed into the Options in the particular tests, and sometimes, like for the `ChainstateManager`, they are ignored. This is down to the mixed usage of both the `m_args`, which creates a fresh, default `ArgsManager` for each test, and the `m_node.args`, which parses all the passed in arguments. 

This was brought up previously in this PR in https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1183527429, where based on the related discussion in https://github.com/bitcoin/bitcoin/issues/25055 and https://github.com/bitcoin/bitcoin/pull/21244#discussion_r615307465 I decided to keep the current behavior. If I understand the current problem correctly, it is not ideal that the `m_node.args` points to the global mutable `gArgs` instead of re-creating the arguments cleanly on each test run, while `m_args` is recreated on each run, but does not read the extra arguments. 

Based on this, I am not sure how to proceed with this on this PR, since I do think that the current approach is broken. I could add a commit making the argument handling consistent for all the `Options`, at least that would make it probably less broken than it is now. What do you think?

mzumsande replied:
Yes, no need to do change it in this PR.

TheCharlatan commented:
Rebased 8f94f059b3af5ecaf175a95389ba5e73b724203b -> 886a473fc48f2c7d67436b5d9ac5643cd007e27f ([removeBlockstorageArgs_23](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_23) -> [removeBlockstorageArgs_24](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_24), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_23..removeBlockstorageArgs_24))
* Fixed conflicts with https://github.com/bitcoin/bitcoin/pull/27573

ryanofsky commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -39,12 +39,14 @@ std::list<const CZMQAbstractNotifier*> CZMQNotificationInterface::GetActiveNotif
     factories["pubhashblock"] = CZMQAbstractNotifier::Create<CZMQPublishHashBlockNotifier>;
     factories["pubhashtx"] = CZMQAbstractNotifier::Create<CZMQPublishHashTransactionNotifier>;
-    factories["pubrawblock"] = CZMQAbstractNotifier::Create<CZMQPublishRawBlockNotifier>;
+    factories["pubrawblock"] = [&get_block_by_index]() -> std::unique_ptr<CZMQAbstractNotifier> {
```
In commit "zmq: Pass lambda to zmq's ZMQPublishRawBlockNotifier" (6247b3c5e3c9e70097243a85b3d883b279a473cc)

It looks like the lambda is taking a reference to the local `get_block_by_index` variable here, which doesn't seem right, since the variable will go out of scope before the lambda is called. So I think `&` should be dropped here.

Not totally sure though, since I would expect there to be problems in CI if there was a bug.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1190097934

Yeah sorry, I saw the "factories" map and assumed the factories would be long-lived, not destroyed before the function returns. This code is pretty strange, because it creates a map and then never looks anything up in the map. Could be cleaned up later, though. Probably a more sane way to write it would be to drop the map, change the outer for loop into an `add_notifiers` lambda, and do

```c++
add_notifiers("pubhashblock", CZMQAbstractNotifier::Create<CZMQPublishHashBlockNotifier>);
add_notifiers("pubhashtx", CZMQAbstractNotifier::Create<CZMQPublishHashTransactionNotifier>);
...
```

In any case, agree the current code is safe

TheCharlatan commented:
Updated 886a473fc48f2c7d67436b5d9ac5643cd007e27f -> e3311d649d9785ca7bc3a29ad7470c7cd69951bb ([removeBlockstorageArgs_24](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_24) -> [removeBlockstorageArgs_25](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_25), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_24..removeBlockstorageArgs_25))
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1190097934), moving the lambda instead of borrowing.

ryanofsky reviewed (approved):
Code review ACK 886a473fc48f2c7d67436b5d9ac5643cd007e27f

re: https://github.com/bitcoin/bitcoin/pull/27125#issuecomment-1542522098

>     * Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1190097934), moving the lambda instead of borrowing.

Actually I think my suggestion won't work. There can be multiple notifiers, so the function needs to be copied into all of them not moved into the first one. ACKed previous version of the PR 886a473fc48f2c7d67436b5d9ac5643cd007e27f though, so would suggest reverting.

fanquake commented:
https://github.com/bitcoin/bitcoin/pull/27125/checks?check_run_id=13380672723
```bash
clang-tidy-16 -p=/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu -quiet /tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/zmq/zmqnotificationinterface.cpp
zmq/zmqnotificationinterface.cpp:48:62: error: std::move of the const variable 'gb' has no effect; remove std::move() or make the variable non-const [performance-move-const-arg,-warnings-as-errors]
        return std::make_unique<CZMQPublishRawBlockNotifier>(std::move(gb));
                                                             ^~~~~~~~~~  ~
```

TheCharlatan commented:
Updated 87479b2c3a6e5044baf41343e5287ff4bef0547a -> 5ff63a09a9edd1204b2cc56cf6f48a44adab7bb3 ([removeBlockstorageArgs_26](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_27) -> [removeBlockstorageArgs_27](https://github.com/TheCharlatan/bitcoin/tree/removeBlockstorageArgs_27), [compare](https://github.com/TheCharlatan/bitcoin/compare/removeBlockstorageArgs_26..removeBlockstorageArgs_27))
* Reverted previous change as suggested by @ryanofsky in [comment](https://github.com/bitcoin/bitcoin/pull/27125#pullrequestreview-1420938880)
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1190113760), correcting the commit message.
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27125#discussion_r1190102588), moving the lambda in the constructor.

ryanofsky reviewed (approved):
Code review ACK 5ff63a09a9edd1204b2cc56cf6f48a44adab7bb3. Since last ACK just added std::move and fixed commit title. Sorry for the noise!

mzumsande commented on src/node/transaction.h:
```diff
@@ -53,11 +54,10 @@ static const CFeeRate DEFAULT_MAX_RAW_TX_FEE_RATE{COIN / 10};
  * @param[in]  mempool         If provided, check mempool for tx
  * @param[in]  hash            The txid
- * @param[in]  consensusParams The params
  * @param[out] hashBlock       The block hash, if the tx was found via -txindex or block_index
```
nit (if you repush): blockman should be added to the doc, also it would be nice to keep the out param at the end according to dev notes.

mzumsande reviewed (commented):
Code Review ACK 5ff63a09a9edd1204b2cc56cf6f48a44adab7bb3
